Piecewise: _eval_interval and _eval_integral overhaul

smichr opened the pull request:
Piecewise changes
-  integration hints are passed along during piecewise integration
-  indefinite integration now returns a continuous result; the old result (integration of pieces) is now called by method `piecewise_integrate`
- _sort_expr_cond has been removed and replaced with _abe which gives unsorted intervals and expressions generated by the Piecewise function; there is not a 1:1 correspondence between the items returned and the Piecewise arguments; the errors associated with encountering unexpected logical functions has been reduced

- [x] check to see that the collapse-code is covered by the added tests
- [x] check to see which of these are still valid with the minimal changes now
- [x] cf https://groups.google.com/forum/#!topic/sympy/_F2FzA0CYZw

closes #11045
closes #12221
closes #7305
closes #12674 
closes #11922 
closes #10137
closes #5227
closes #12557 
closes #6900
closes #10122
closes #4313
closes #9885
closes #12976

- [x] fix piecewise_fold to maintain arg order for arbitrary args by using cartes on the piecwise args or (a, True) for non-piecewise args
- [x] leave instantiation simple wrt filtering out duplicates and let that be handled by `smooth` otherwise if a `not(c)` is seen in or_seen, that is a redundant condition, too, e.g. `(1,x>0),(2,And(x<=0,x>-1)) -> (1,x>0),(2,x>-1)`
- [x] add tests from #8919
- [x] check that pw defined over range of integration but with no default value works
- [x] when this is committed, make a note [here](http://stackoverflow.com/questions/43852159/wrong-result-when-integrating-piecewise-function-in-sympy/43854090#43854090) because that integral gives the correct answer now.
- [x] make sure that new `piecewise_integrate` and `_eval_integral` are ok without deprecation
- [x] ~~name `_handle_irel` as `piecewise_unfold(x)`? Or can this function be removed by making `abe` emit (a, b, e, aux_conditions)?~~
- [x] replace `_sort_expr_cond` with a more robust integrator of piecewise functions; this is currently being tested.
- [x] break out `_msp`
- [x] break out `piecewise_fold` since this is a small (vital) part that is easy to stand separately
- [x] unify holes handling:
```
>>> Piecewise((a,Eq(1,x)),(b,And(x>0,x<3))).integrate((x,a,b))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy\core\expr.py", line 3082, in integrate
    return integrate(self, *args, **kwargs)
  File "sympy\integrals\integrals.py", line 1311, in integrate
    risch=risch, manual=manual)
  File "sympy\integrals\integrals.py", line 569, in doit
    evalued = Add(*others)._eval_interval(x, a, b)
  File "sympy\functions\elementary\piecewise.py", line 537, in _eval_interval
    raise ValueError("Can't integrate across undefined region.")
ValueError: Can't integrate across undefined region.
>>> Piecewise((a,Eq(1,x)),(b,And(x>0,x<3))).integrate(x)
nan

Results and the range in which they are valid could be given 
instead of nan, but perhaps an error as in the definite case 
would be better for the indefinite case instead of `nan`.

But in this case we get a partial answer because the integral is defined for x > -oo
up to 2:

>>> Piecewise((1, x < 2)).integrate(x)
Piecewise((x, x < 2), (nan, True))

```
- [x] is it necessary to allow for lists to be expressions in Piecewise to be consistent with solve  output? 
A: they can already be args in master. Only the _subs method had to be modified to deal with them.
```
in master

>>> Piecewise(([1,2,3],x>1))
Piecewise(([1, 2, 3], x > 1))
>>> _.subs(x,2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy\core\basic.py", line 915, in subs
    rv = rv._subs(old, new, **kwargs)
  File "sympy\core\cache.py", line 93, in wrapper
    retval = cfunc(*args, **kwargs)
  File "sympy\core\compatibility.py", line 794, in wrapper
    return user_function(*args, **kwds)
  File "sympy\core\basic.py", line 1027, in _subs
    rv = self._eval_subs(old, new)
  File "sympy\functions\elementary\piecewise.py", line 455, in _eval_subs
    e = e._subs(old, new)
AttributeError: 'list' object has no attribute '_subs'
```
- [x] check to see that this was tested
- [x] make Relational.canonical simpler
- [x] find out why test_transforms.py line 300 and 473 are failing
- [x] make sorting of roots a different PR
- [x] separate out Piecewise handling for solve and make new PR; don't allow list as an arg in Piecewise [recall this conversation](https://github.com/sympy/sympy/pull/12920#discussion_r126713297)
- [x] debug tests at and below the following
test_piecewise.py", line 671, in test_issue_11045
test_transforms.py", line 530, in test_issue_8368_7173
test_transforms.py", line 468, in test_laplace_transform

- [x] check on solving of integral in #13140
- [x] check if ` lower, upper = lower, Max(lower, upper) ` should be `...= Min(lower, upper), Max...`.
    A: no, it fails then
- [x] change abei to intervals?
- [x] if `rv.rel_op == '!=':  rv = Or(sym < rv.rhs, sym > rv.rhs)` will fail if rhs is `I`; perhaps `sym<oo` should be used instead
- [x] extract the inequalities modifications and the formal power series modification
- [x] try to extract the changes to integrals.py
- [x] get rid of C3 in test_ode in the line `x, y, z, C3 = symbols('x, y, z, C3', function=True)`

~~- [ ] break out the Piecewise.eval changes~~

smichr commented on sympy/functions/elementary/piecewise.py:
```diff
@@ -223,6 +224,9 @@ def _eval_interval(self, sym, a, b):
             values = []
             intervals = self._sort_expr_cond(
                 sym, S.NegativeInfinity, S.Infinity)
+            if len(intervals) == 1 and intervals[0][-1] == self:
```
Note to self: Some of the failing Piecewise integration tests (from issues) involve intervals in which the expressions are Piecewise, so perhaps this is just a special case of that and needs to be handled by an evaluation of that piecewise like is handled below for this special case.

smichr commented on sympy/integrals/tests/test_transforms.py:
```diff
@@ -115,22 +114,22 @@ def test_mellin_transform():
     mt = MT((1 - x)**(beta - 1)*Heaviside(1 - x)
             + a*(x - 1)**(beta - 1)*Heaviside(x - 1), x, s)
-    assert mt[1], mt[2] == ((0, -re(beta) + 1), True)
+    assert mt[1], mt[2] == ((0, -re(beta) + 1), re(beta) > 0)
```
This needs to be checked to see if the change from `True` to `re(beta) > 0` is correct

maurogaravello replied:
The convergence conditions seem correct.

smichr replied:
Thanks (and to @jksuom ). This was a badly formed test that would have passed unless `mt[1]` was false.

smichr commented:
I would really appreciate another set of eyes taking a look at these changes. Several issues have been resolved. There are still issues with integrating piecewise when there are symbols in the limits of integration or in the piecewise definition, but that is a whole other ball of yarn. This is a good place to stop and commit what is present.

jksuom commented on sympy/core/relational.py:
```diff
@@ -109,13 +119,16 @@ def canonical(self):
         r = self
         if r.func in (Ge, Gt):
-            r = r.reversed
+            if _coeff_isneg(r.rhs):
```
I feel that Gt should be converted to Lt only under more special conditions. So I would personally prefer conditions like `x > a + b`, `x > f(a)` and even `x > a` when `x` is variable and `a` is constant. I think that it is more informative to have the variable on the left hand side. (That is the order typically used in verbal communication: "We have `<some property or expression of x>`when x is greater/smaller than `<something>`".)

asmeurer replied:
It's complicated, because something like 1 < x < 4, which has to be printed (at least with the str printer) using &&, is easier read as (1 < x) && (x < 4) than (x > 1) && (x < 4).  

My take here would be to put effort in the pretty printers (`pprint` and `latex`) to print things better, and not worry as much about the expression form itself.

smichr commented on sympy/core/relational.py:
```diff
@@ -109,13 +119,16 @@ def canonical(self):
         r = self
         if r.func in (Ge, Gt):
-            r = r.reversed
+            if _coeff_isneg(r.rhs):
```
I thought this modification was necessary to avoid an infinite recursion that I was getting where  I was getting alternation between `x**2 > -y` and `y > -x**2`. But I see in master that the canonical form is the latter (as we would expect).

As to your sentiments regarding what looks right, I tried to respect those with the former introduction of the canonical method:

```
>>> from sympy.core.relational import _canonical as f  # new function to make expr canonical
>>> f(x>a)
a < x
>>> f(x>a+b)
x > a + b
>>> f(x>a+1)
x > a + 1
>>> f(Or(x>b,x<a))
(b < x) | (x < a)
>>> f(Or(x<a,x>b))
(b < x) | (x < a)
>>> f(And(a<x,x<b))
(a < x) & (x < b)
```

I'll remove this modification and try better locate the source of the recursion.

jksuom replied:
I realize that it is not easy to devise a good plan for canonicalization of inequalities. Right now I can only propose simple (and obviously inadequate) rules such as not changing the direction if the LHS is a single symbol. (It would probably be better if one could tell variable and constant symbols apart.)

smichr commented:
@jksuom , I think I have a better fix in place for canonical.

smichr commented:
This is taking a break. Please feel free to leave comments or to find a more concise solution for the issues. I don't know how to do this other than to use the permutations.

jksuom commented:
I am afraid that the use of permutations will bring about an element of unpredictability. I think that it is important to preserve the order of bounds as they appear in `abe`, the list whose elements consist of an interval `(a, b)` and the associated expression `e`. For each element `(a, b, e)`, the interval `(a, b)` should be reduced by removing those parts that belong to an interval occurring earlier in the list. For example, for `Piecewise((0, x < a), (1, x > b), (2, True))` we have
    `abe = [(-oo, a, 0), (b, oo, 1), (-oo, oo, 2)]`,
and after reducing the intervals from left to right we get
    `abe = [(-oo, a, 0), (Max(a, b), oo, 1), (a, Max(a, b), 2)]`.
This list can now be rearranged in growing order before integration:
    `abe = [(-oo, a, 0), (a, Max(a, b), 2),  (Max(a, b), oo, 1)]`.

Sometimes an interval may be split into several pieces when pieces of earlier intervals are removed. In that case, each one should be added as a new element to `abe`. (And they will be separated when the intervals are rearranged.)

smichr commented:
> I am afraid that the use of permutations will bring about an element of unpredictability

Yes, I am aware of the arbitrary nature of the output as it caused me some difficulty in writing tests. I believe I have that resolved that so the result is now deterministic. I did this by ordering the symbols before they enter the permutation. The results that are obtained are arbitrary but should always be the same (unless the permutation engine changes). Although I would like to use something as you indicate above, I'm not sure how to convert something like this (with arrows indicating unbounded intervals with higher priority intervals nearer the top) into a Min/Max ordering *for each interval*. 
```
a------->
<-------b
c-------d
<------->

or 

a------->
c-------d
<-------b
<------->

or

c-------d
a------->
<-------b
<------->
```

e.g. the first point encountered will be Min(a,b,c), but what comes next and what interval will it be in? (That's the crux of my issue, I think.)  Certainly for the univariate case (i.e. all fixed boundaries) this can be done, but can it be done in general?

jksuom commented:
> what comes next and what interval will it be in?

I think that these questions can be answered if the order of the modified (max/min) endpoints is noted down along the computations.

In the first example above, the first interval `(a, oo)` is taken unmodified and its endpoints are saved in a list in the given order. The next interval will be the intersection of `(-oo, b)` with the complement of the first interval. That will be `(-oo, b')` where `b' = Min(a, b)` for brevity, and we note that `b' <= a`.

The third interval `(c, d)` will be intersected with the complements `(-oo, a)` and `(b', oo)` of the previous intervals. We first get `(c', d')` where `c' = Min(a, c)` and `d' = Min(a, d)`. (If `a <= c`, then this becomes `(a, a)` which is empty.) At this point we have `c' <= d' <= a`. Intersecting with `(b', oo)` we get `(c", d")` where `c" = Max(b', c')` and `d" = Max(b', d')`, and the  ordering will be `b' <= c" <= d" <= a`.

The intervals carrying the default case can be found similarly by taking the complements of the previous intervals in `(-oo, oo)`. In general there will be two intervals, either of which, or both, may be empty, namely `(b', c")` and `(d", a)`.

smichr commented:
TIP [thought in progress]

What I have done so far (given interval [a,b] and [c,d] (with [a.b] having higher precedence than [c,d]) is to define what regions the behavior in [a,b] applies, the behavior in [c,d] applied and the region in which neither applies. So if I am thinking correctly about this, 

- the region in which [a,b] applies is the intersection of the two regions
- everything else is the complement, C; 
- the region in which [c,d] applies is the intersection of C and [c,d]

jksuom commented:
> the region in which [a,b] applies is the intersection of the two regions

Do you mean the intersection of [a, oo) and (-oo, b]?

smichr commented:
> Do you mean the intersection of [a, oo) and (-oo, b]?

Let's start over. This is my current thinking about how to solve this:

```
done = []
undone = [(-oo, oo)]
pieces = [(a, oo), (-oo, b), (c, d)]
for p in pieces:
    done.append([intersection(p, u) for u in undone])
    _undone = []
    for co in complement(*done[-1]):
        for u in undone:
            _undone.append(intersection(u, co))
    undone = _undone
```
which gives the following list of intervals "done" after each step
```
[(a, oo)]
[(-oo, Min(a, b))]
[(Max(c, Min(a, b)), Max(c, Min(a, b), Min(a, d)))]
```
This differs from your result.The difference is that I don't intersect with *each* of the previous complements, I intersect with the cumulative complement -- or the intervals left undone after processing an interval. The above code gives two intervals that might be undone:

```
(Min(a, b), Min(a, Max(c, Min(a, b))))
(Max(c, Min(a, b), Min(a, d)), Max(a, c, Min(a, b), Min(a, d)))
```

With `reps = {a:4,b:0,c:1,d:2}` this gives:

```
>>> [list(Tuple(*i).subs(reps)) for i in done]
[[(4, oo)], [(-oo, 0)], [(1, 2)]]
>>> Tuple(*undone).subs(reps)
((0, 1), (2, 4))
```

The basic tools I am using are routines to calculate the intersection of two intervals, the complement of a list of (ordered?) intervals and the possible holes between two intervals. I suspect there is a better way to calculate what is undone; three general intervals gives 42 possible undone regions (many of which will reduce to null intervals when values are substituted). 

```
def intersection(A, B):
    (a, b) = A
    (c, d) = B
    if 0:
        (a, b) = (Min(a, b), Max(a, b))
        (c, d) = (Min(c, d), Max(c, d))
    if a is -oo and b is oo:
        rv = (c, d)
    elif c is -oo and d is oo:
        rv = (a, b)
    elif b is oo and d is oo:
        # a,oo; c,oo
        rv = (Max(a, c), oo)
    elif a is -oo and c is -oo:
        # -oo,b;  -oo,d
        rv = (-oo, Min(b, d))
    elif a is -oo and d is oo:
        # -oo,b; c,oo
        rv = (Min(b, c), b)
    elif b is oo and c is -oo:
        # a,oo; -oo,d
        rv = (Min(a, d), d)  # Min(a, d), Min(d, Max(a, d))
    elif d is oo:
        # a,b; c,oo
        rv = (Max(a, c), Max(b, c))  # or (Min(b, Max(a, c)), b)
    elif c is -oo:
        # a,b; -oo,d
        rv = (Min(a, d), Min(b, d))  # or (Max(a, Min(b, d)), a)
    elif b is oo:
        # a,oo; c,d
        rv = (Max(a, c), Max(a, d))  # or (Min(d, Max(c, a)), d)
    elif a is -oo:
        # -oo,b; c,d
        rv = (Min(c, b), Min(d, b))  # or (Max(c, Min(d, b)), c)
    else:
        # a,b; c,d
        rv = (Max(a, c), Max(a, c, Min(b, d)))
    return rv

def complement(*iv):
    l = r = -oo
    got = []
    for a, b in iv:
        if a is -oo:
            l = b
        elif b is oo:
            got.append((r, a))
            break
        else:
            got.append((l, a))
            l = b
    if l is not oo and l is not -oo:
        got.append((l,oo))
    return got or [(-oo,oo)]

def holes(A, B, clip=True):
    (a, b) = A
    (c, d) = B
    assert not b is -oo
    assert not a is oo
    assert not d is -oo
    assert not c is oo
    # 2 infinities in one
    if a is -oo and b is oo:
        rv = []
    elif c is -oo and d is oo:
        rv = []
    else:
        # 1 infinity in each
        if b is oo and d is oo:
            # a,oo; c,oo
            rv = [(-oo, Min(a, c))]
        elif a is -oo and c is -oo:
            # -oo,b; -oo,d
            rv = [(Max(b, d), oo)]
        elif a is -oo and d is oo:
            # -oo,b; c,oo
            rv = [(b, Max(b, c))]
        elif b is oo and c is -oo:
            # a,oo; -oo,d
            rv = [(Min(a, d), a)]
        else:
            # an infinity in one of them
            if c is -oo or d is oo:
                (a, b, c, d) = (c, d, a, b)
            if b is oo:
                # a,oo; c,d
                if clip:
                    rv = [(d, Max(a, d))]
                else:
                    rv = [(-oo, Min(a, c)), (d, Max(a, d))]
            elif a is -oo:
                # -oo,b; c,d
                if clip:
                    rv = [(b, Min(Max(b, c), Max(b, d)))]
                else:
                    rv = [
                        (b, Min(Max(b, c), Max(b, d))),
                        (Max(b, d), oo)]
            else:

              # no infinities
              # a,b; c,d
                if clip:
                    rv = [(Max(Min(a, d), Min(b, c)), Min(Max(a, d), Max(b, c)))]
                else:
                    rv = [
                        (-oo, Min(a, c)),
                        (Max(Min(a, d), Min(b, c)), Min(Max(a, d), Max(b, c))),
                        (Max(b, d), oo)]
    return [(a, b) for (a, b) in rv if a != b]
```

jksuom commented:
> This differs from your result.

This is probably not caused by the difference between the two methods of intersecting (sequence of complements vs. complement of union) but by the inherent ambiguity in the form of intersections. 
That is visible already in the intersection of  `(-oo, b)`  and  `(c, oo)`  which can written  `(Min(b, c), b)`  or dually  `(c, Max(b, c))`.
Correspondingly, the intersection of `(a, b)` and `(c, d)`, which is equal to the intersection of `(-oo, Min(b, d))` and `(Max(a, c), oo)` can be either `(Min(b, d, Max(a, c)), Min(b, d)` or `(Max(a, c), Max(Min(b, d), a, c)`.

I did choose to intersect with complements of intervals since those consist of (one or two) unbounded intervals and it seemed that it was easier to the find the proper order of the new endpoints that way. But I think your method is equally valid.

smichr commented:
> choose to intersect with complements of intervals

The problem I run into is that I end up generating duplicates since if [c,d] is intersected with the first and second complement, there are intersections with the first that would have been removed if I intersected with the union of the first and second.  Here is how I understand what you said to work:

```
pieces = [(a, oo), (-oo, b), (c, d)]
done = [pieces[0]]
comps = complement(pieces[0])
for p in pieces:
    more = []
    for c in comps:
        new = intersection(p, c)
        more.extend(complement(new))
        done.append(new)
    comps.extend(more)

>>> reps = {a:4,b:0,c:1,d:2}
>>> done
[(a, oo), (a, a), (-oo, Min(a, b)), (-oo, Min(a, b)), (Min(a, b), b), (Min(a, c), Min(a, d)), (Min(a, c), Min(a, d)), (Max(a, c), Max(a, d)), (Max(c, Min(a, b)), Max(d, Min(a, b))), (Max(c, Min(a, b)), Max(d, Min(a, b))), (Min(a, b, c), Min(a, b, d)), (Max(b, c), Max(b, d))]

>>> set(done)
{(Min(a, c), Min(a, d)), (Min(a, b), b), (Max(a, c), Max(a, d)), (Max(b, c), Max(b, d)), (a, oo), (a, a), (Max(c, Min(a, b)), Max(d, Min(a, b))), (Min(a, b, c), Min(a, b, d)), (-oo, Min(a, b))}

>>> Tuple(*set(done)).subs(reps)
((1, 2), (0, 0), (4, 4), (1, 2), (4, oo), (4, 4), (1, 2), (0, 0), (-oo, 0))
```

jksuom commented:
I'll try to describe a possible algorithm. First a list, `pieces`, of possibly overlapping intervals is prepared from the arguments of Piecewise (preserving the priority order). There are expressions associated with the intervals but those are ignored here.

The algorithm proceeds by producing a sequence of new `done` lists of intervals in increasing order, one for each `p` in `pieces`, such that the union of the intervals is equal to the union of the intervals in `pieces` up to and including `p`. The first nonempty `done` list will contain only the first interval in `pieces`.

Given an intermediate `done` list and the next interval `p` in `pieces`, the updated `newdone` list is constructed by adding subintervals of `p` to gaps of `done`. A gap is a nonempty interval between two consecutive intervals in `done`. (If the first interval `(a, b)` does not start from `-oo`, then the first gap is `(-oo, a)`.)

For each gap `(a, b)`, its intersection with `p` is considered. To this end two auxiliary functions are defined. The first function is the retraction to `(a, b)`. The value of `retract(x, a, b)` is defined by
`a if x < a else b if x > b else x`,
and it can be written as `Min(Max(x, a), b)` or `Max(Min(x, b), a)`.

The second function, `restrict(p, a, b)` returns the interval
`(retract(p.left, a, b), retract(p.right, a, b)`.
This is a subinterval of the gap `(a, b)` and it will be entered into `newdone` in the position of the gap unless it is empty, i.e., its endpoints are equal. (That will happen when it can be seen that both endpoints of `p` are either less than `a` or greater than `b`.)

Thus `newdone` can be constructed from an empty list by scanning `done` and appending alternately an interval of `done` or (potentially) a restriction of `p` to a gap.

smichr commented:
> For each gap (a, b), its intersection with p is considered. To this end two auxiliary functions are defined

I already have the routine to analyze two intervals and produce an ordered sequence of intervals that are covered by the first, second or neither of the intervals. What I didn't know how to do was maintain that order when bringing in the 3rd interval. Replacing a gap with the third interval *after restricting it* is the key. That's the idea I was missing. Thanks for the help.

smichr commented:
@jksuom , splitting a 3-arg piecewise function into intervals with Min and Max is working. Here is the result for intervals 0, 1 and 2 in the ranges [a,b], [c,d], and [e,f], respectively:

```
(-oo, Min(a, c, e), -1)
(Min(a, c, e), Min(a, c, f), 2)
(Min(a, c, f), Max(Min(a, c), Min(a, c, f)), -1)
(Min(a, c), Min(a, d), 1)
(Min(a, d, Max(e, Min(a, d))), Min(a, Max(e, Min(a, d))), -1)
(Min(a, Max(e, Min(a, d))), Min(a, Max(f, Min(a, d))), 2)
(Min(a, Max(f, Min(a, d))), Max(a, Min(a, Max(f, Min(a, d)))), -1)
(a, b, 0)
(Min(b, Max(b, c), Max(b, e)), Min(Max(b, c), Max(b, e)), -1)
(Min(Max(b, c), Max(b, e)), Min(Max(b, c), Max(b, f)), 2)
(Min(Max(b, c), Max(b, f)), Max(b, c, Min(Max(b, c), Max(b, f))), -1)
(Max(b, c), Max(b, d), 1)
(Min(Max(b, d), Max(b, d, e)), Max(b, d, e), -1)
(Max(b, d, e), Max(b, d, f), 2)
(Max(b, d, f), oo, -1)
```

for values of a-f of 0-6 gives

```
(-oo, 0, -1)
(0, 0, 2)
(0, 0, -1)
(0, 0, 1)
(0, 0, -1)
(0, 0, 2)
(0, 0, -1)
(0, 1, 0)
(1, 2, -1)
(2, 2, 2)
(2, 2, -1)
(2, 3, 1)
(3, 4, -1)
(4, 5, 2)
(5, oo, -1)
```

jksuom commented:
I'm glad to see that this can be used as a basis for the treatment of piecewise functions. The expression for three generic intervals (plus a default) looks quite complicated but the simplified form is (almost) understandable. (Another reason for making the use of absorption law automatic.) In practical situations all the endpoints are rarely variable, and simplifications will occur along the way before reaching the final expression.

smichr commented:
> absorption law

I did this simplification by replacing Min/Max with And/Or, doing simplification, and returning back to Min/Max. There are some absorptions for which I don't know the rule (and I'm not sure it is even an absorption rule). e.g. 

```
Min(Max(b, c), Max(b, e)) -> Max(b, Min(c, e))
```

jksuom commented:
> Min(Max(b, c), Max(b, e)) -> Max(b, Min(c, e))

This is an example of lattice distribution law. It is fully analogous to the arithmetic distribution law

    Add(Mul(b, c), Mul(b, e)) == Mul(b, Add(c, e))

There is another lattice distribution law with `Max` and `Min` interchanged:

    Max(Min(b, c), Min(b, d)) == Min(b, Max(c, d))

It is said that `Max` and `Min` are distributive with respect to each other.
The situation is asymmetric in arithmetic. Multiplication is distributive with respect to addition but not vice versa.

smichr commented:
The MinMax results are much more compact than what I had yet attained with the Relational approach. I have (mostly for personal reference) a comparison of the general integration between symbolic limits for a variety of Piecewise functions:
```
>>> s = Piecewise((A,x>=a,x<=b),(B,True)).integrate((x,c,d))
>>> ll=Piecewise((A, x<=a), (B, x<=b),(C,True)).integrate((x,e,f))
>>> lr=Piecewise((A, x<=a), (B, x>=b),(C,True)).integrate((x,e,f))
>>> rr=Piecewise((A, x>=a), (B, x>=b),(C,True)).integrate((x,e,f))
>>> rl=Piecewise((A, x>=a), (B, x<=b),(C,True)).integrate((x,e,f))
>>> sr=Piecewise((A, (a <= x) & (x <= b)), (B, x>=c),(C,True)).integrate((x,e,f))
>>> len(str(sr))  # the min/max result is 457 chars long
136507
```
[it is at this gist](https://gist.github.com/smichr/20d35e07715a6b83cdcc62fea51623ff)

smichr commented:
:memo: Look into what is going on with

```
>>> p=Piecewise((A,And(x<1,x>0)),(B,And(x>y,x<2)),(C, True))
>>> integrate(p,x)
Piecewise((C*x, x <= Min(1, y)))
>>> p._abei(x)
[(0, 1, A, 0), (y, Max(2, y), B, 1), (-oo, oo, C, 2)]
>>> integrate(p,x)
Piecewise((C*x, x <= Min(1, y)))
>>> integrate(p,(x,-oo,oo))
C*Min(1, y) + oo*sign(C)
>>> p=Piecewise((A,And(x<1,x>0)),(C, True))
>>> integrate(p,x)
Piecewise((C*x, x <= 1))
```

smichr commented on sympy/core/tests/test_relational.py:
```diff
@@ -1,12 +1,12 @@
                                    GreaterThan, LessThan, StrictGreaterThan,
                                    StrictLessThan, Rel, Eq, Lt, Le,
-                                   Gt, Ge, Ne)
+                                   Gt, Ge, Ne, _canonical)
```
add in tests for `_canonical`

smichr commented on sympy/functions/elementary/miscellaneous.py:
```diff
@@ -337,7 +337,6 @@ def __new__(cls, *args, **assumptions):
 
         # second filter
         # variant I: remove ones which can be removed
-        # args = cls._collapse_arguments(set(_args), **assumptions)
```
Though this function doesn't exist, perhaps this should still be left as a reminder for when it did.

smichr commented on sympy/functions/elementary/miscellaneous.py:
```diff
@@ -354,6 +353,32 @@ def __new__(cls, *args, **assumptions):
             obj._argset = _args
             return obj
 
+    def _eval_simplify(self, ratio, measure):
```
this could be a separate PR

smichr commented on sympy/integrals/integrals.py:
```diff
@@ -764,7 +780,7 @@ def _eval_integral(self, f, x, meijerg=None, risch=None, manual=None,
 
         # Piecewise antiderivatives need to call special integrate.
         if f.func is Piecewise:
-            return f._eval_integral(x)
```
Is this the source of difference in results? Should the `_eval_integral` be used?

smichr commented on sympy/integrals/tests/test_prde.py:
```diff
@@ -306,3 +306,16 @@ def test_parametric_log_deriv():
     assert parametric_log_deriv_heu(Poly(5*t**2 + t - 6, t), Poly(2*x*t**2, t),
     Poly(-1, t), Poly(x*t**2, t), DE) == \
         (2, 6, t*x**5)
+
```
delete this -- it was relocated by #12743

smichr commented on sympy/solvers/solvers.py:
```diff
@@ -1122,6 +1122,7 @@ def _do_dict(solution):
             solution and
             ordered_symbols and
             type(solution) is not dict and
+            type(solution) is not Piecewise and
```
separate out Piecewise handling for solve and make new PR; don't allow list as an arg in Piecewise [recall this conversation](https://github.com/sympy/sympy/pull/12920#discussion_r126713297)

smichr commented:
- [ ] find out why test_transforms.py line 300 and 473 are failing.

smichr commented on sympy/solvers/solvers.py:
```diff
@@ -15,7 +15,7 @@
 from __future__ import print_function, division
```
- [ ] make this a separate PR

smichr replied:
- [ ] make sorting of roots a different PR

asmeurer commented on sympy/integrals/integrals.py:
```diff
@@ -428,67 +428,83 @@ def doit(self, **hints):
+                    else:
+                        meijerg1 = False
+                        '''
+If the special meijerg code did not succeed in finding a definite
```
Why are these strings here? If this is intended to be a comment, it should be a comment, and indented along with the code block.

asmeurer commented:
As long as you're refactoring Piecewise and Integral, it would be helpful to remove the Piecewise specific code from the ExprWithLimits superclass into more specific place (this was causing some issues for @leosartaj because fps objects don't work with piecewise_fold).

asmeurer commented on sympy/functions/elementary/piecewise.py:
```diff
@@ -195,242 +310,422 @@ def _eval_derivative(self, x):
+                    sym in rv.free_symbols:
+                if rv.args[0] != sym:
+                    raise NotImplementedError(filldedent('''
+Unable to solve relational %s for %s.''' % (r, sym)))
```
Indent this along with the rest of the code (isn't that the whole point of using filldedent?)

smichr replied:
Part of the point is also that you don't have to worry about where it is indented to. The indentation at this point in the code would require me to write something like

```
Unable to solve relational %s for %s.''' % (r,
sym)))
```

I am `random()/10` for favoring what I have, so if you prefer, I'll change it.

asmeurer commented on sympy/functions/elementary/piecewise.py:
```diff
@@ -195,242 +310,422 @@ def _eval_derivative(self, x):
+            sum += abei[i][-2]._eval_interval(x, a, b)
+        return _mmsimp(sum)
+
+    def _abei(self, sym):
```
Can a more descriptive name be used here?

smichr replied:
yes...I've still got a nagging todo to return intervals that are open or closed instead of simply a and b.

asmeurer commented on sympy/functions/elementary/piecewise.py:
```diff
@@ -195,242 +310,422 @@ def _eval_derivative(self, x):
+        return _mmsimp(sum)
+
+    def _abei(self, sym):
+        """Return (a, b, e, i) where a and b are the lower and
```
The above code suggests it actually returns a list of `(a, b, e, i)`.

smichr replied:
fixed

smichr commented:
> this was causing some issues for @leosartaj because fps objects don't work with piecewise_fold

I searched issues for fps and only found one but didn't quickly see the problem with piecewise. Can you point me to a discussion pertaining to this?

asmeurer commented:
The discussion happened in person at SciPy, so it wasn't written down. The pull request is https://github.com/sympy/sympy/pull/12976/files#diff-2b7bb119e0248507ad744c428ea9e7ffR87. That line is the workaround that @leosartaj used to work around the issue. I believe if you remove it and run the test that was added you can see what the problem is.

smichr commented:
> it wasn't written down. The pull request is

That was helpful, thanks. I've taken a quick look. I'll have to dig into it a little more but I'll see what I can do.

smichr commented:
> remove the Piecewise specific code from the ExprWithLimits superclass into more specific place

Since it is needed for integration, it seems that putting it there makes sense. You can see the most recent commit for my attempt at doing so.

smichr commented on sympy/integrals/integrals.py:
```diff
@@ -497,10 +497,8 @@ def try_meijerg(function, xab):
-                        conds=conds)
-                    if antideriv is None and meijerg1 is True:
+                        function, xab[0], **eval_kwargs)
+                    if antideriv is None and meijerg is True:
```
changed meijerg1 to meigerg; the only way meijerg1 could be True is if meijerg were True because the intervening code only sets meijerg1 to False if the integration failed.

smichr commented on sympy/integrals/integrals.py:
```diff
@@ -476,15 +475,16 @@ def try_meijerg(function, xab):
 
                 meijerg1 = meijerg
-                if (len(xab) == 3 and xab[1].is_real and xab[2].is_real
+                # XXX shouldn't the following be attempted only if meijerg is not False?
```
remove the XXX if the tests pass

smichr commented on sympy/integrals/integrals.py:
```diff
@@ -399,16 +401,22 @@ def doit(self, **hints):
+            return function.applyfunc(
+                lambda f: self.func(f, self.limits).doit(**hints))
 
         if isinstance(function, FormalPowerSeries):
```
@leosartaj , does this look right to you? It seems odd/wrong to integrate wrt the first limit if there are multiple integrations...so now (in case this case exists) a NotImplementedError is raised if there is more than one integation, e.g. `integrate(foo, lim1, lim2)`

smichr committed:
overhaul

smichr commented on sympy/solvers/inequalities.py:
```diff
@@ -539,7 +539,7 @@ def valid(x):
                 v = expanded_e.subs(gen, x)
                 try:
-                    r = expr.func(v, 0)
+                    r = expand_mul(expr.func(v, 0))
```
use `v = expanded_e.subs(gen, expand_mul(x))` above instead of using `expand_mul` here

smichr commented on sympy/functions/elementary/piecewise.py:
```diff
@@ -58,7 +62,7 @@ def free_symbols(self):
     @property
     def is_commutative(self):
-        return self.expr.is_commutative
+        return getattr(self.expr, 'is_commutative', False)
```
All args are Basic so this change is not necessary.

smichr committed:
eliminate unneeded routines; expand x

smichr commented:
@jksuom and @asmeurer , do you have any suggestions on how this can move forward? It's a lot to review. What can I do to make the review easier? I've pulled out almost everything that can stand alone. As a last effort I was thinking of moving the eval changes out so it is clear what tests are changing because of that and which are changing because of the integration related changes. I'm not sure if this is worth it. I could really use some advice.

Would it help to add an explanatory docstring in the `_eval_integral` routine to tell how the process is working? I'm not sure if there is a better way to accomplish these goals (although I know from experience documented here that there is a *worse* way!)

asmeurer commented on sympy/functions/elementary/piecewise.py:
```diff
@@ -664,3 +967,222 @@ def piecewise_fold(expr):
+    Min and Max functions:
+
+    >>> _pair((x, 2), (4, 5))
+    {-1: [(-oo, Min(4, x)), (Min(5, x), x), (2, 4), (5, oo)],
```
Couldn't this be simpler, since x has to be <= 2?

smichr replied:
I believe this is not simpler because it aims to be right even if x is greater than 2. If you follow along in the docstring, different values are given for x. I'm not sure of the implications if the assumption is made that x <= 2. It's probably worth revisiting. I'll take a look. One might even be able to trick the mins into resolving by supplying a known displacement to one arg, e.g. if x is assumed to be less than 2 then x could be replaced with `2 - nonnegative` and then `Min(2 - nonnegative, 4)` would resolve to `2 - nonnegative` and backsubstituting nonnegative with `2 - x` would yield x again instead of `Min(x, 4)`.

asmeurer replied:
I'm working on the assumption that if x >= 2 then the interval (x, 2) is empty. In that case you have just -oo -- 4 -- 5 -- oo. I would need to think about this. It's possible I am wrong that it can be simplified. Certainly in a case like (2, x), (4, 5) it cannot be simplified.

smichr committed:
use lo/hi ordering to simplify min/max args

smichr committed:
remove unused import

smichr committed:
edit inline notes and spacing

jksuom commented on sympy/functions/elementary/tests/test_piecewise.py:
```diff
@@ -187,57 +195,122 @@ def test_piecewise_integrate():
+        assert p.subs(r).integrate(lim.subs(r)) == q.subs(r)
+
+
+def test_miejer_bypass():
```
miejer -> meijer

jksuom commented on sympy/functions/elementary/tests/test_piecewise.py:
```diff
@@ -550,6 +687,300 @@ def test_S_srepr_is_identity():
+    assert Piecewise((d1, Eq(x, 2)), (0, True)).integrate(x) == 0  # right
+    # XXX should the following be
+    # Piecewise((Heaviside(0), Eq(x, 1)), (1, x > 1), (0, True))?
+    assert Piecewise((d1, Eq(x, 1)), (0, True)).integrate(x) == 0
```
I think this test should not be included. Although it is possible to say that `d1` is 0 for `x != 1`, its value at 1 cannot be considered as a number. There is no reasonable way of assigning some meaning to `Piecewise((d1, Eq(x, 1)), (0, True))`.

smichr replied:
From my (very limited) understanding of DD, I am not thinking there is a value I can assign other than DD(0) at x=1, but I was thinking of the answer in the inline comment based on the following:

```
d1=DiracDelta(x - 1)
p = Dummy(positive=True)
integrate(d1, (x, -oo, 1 - p))
0
integrate(d1, (x, -oo, 1))
Heaviside(0)
integrate(d1, (x, -oo, 1 +p))
1
```

jksuom replied:
The expression `integrate(d1, (x, -oo, 1))` is also ambiguous since it is does not make clear whether 1 is included in the domain of integration or not. To define integrals of singular measures it is necessary to specify the domain exactly. The integral of `d1` over the open interval `(-oo, 1)` is 0 but over the half-closed interval `(-oo, 1]` it is 1. It should also be noted that the [fundamental theorem of calculus](https://en.wikipedia.org/wiki/Fundamental_theorem_of_calculus#Formal_statements) on the connection between integral and antiderivative is valid for continuous functions only. It cannot be used if the integrand is discontinuous at one of the limits of integration.

smichr replied:
I will delete the test and leave this as a new issue.

jksuom commented:
This looks good to me. I have not inspected everything, but I think this is a big improvement and should be merged.

smichr committed:
align inlines; delete DiracDelta test; miejerg typo

smichr commented:
I am working to see if pairs results can be improved.

jksuom commented:
It could be possible to remove the big `if-elif`-switch by using the last `else`-expression for `br` in all cases. The next for-loop would then remove the redundant intervals.

smichr commented:
> I am working to see if pairs results can be improved.

Hmmm. I tried to use `a = Min(a, b)` and `d = Max(c, d)` in the last expression and used only that expression for br:
```
br = (-oo, -1, Min(a, c), 1, Min(a, d), -1, a, 0, b, -1, 
    Max(b, c), 1, Max(b, d), -1, oo)
became
br = (-oo, -1, Min(Min(a, b), c), 1, Min(Min(a, b), Max(c, d)), -1, Min(a, b), 0, b, -1, 
    Max(b, c), 1, Max(b, Max(c, d)), -1, oo)
```
While this fixes it so that at worst you only get a null range from `_pair` other things break wrt integration of Piecewise in some significant ways which are clearly wrong. So other than @jksuom 's suggestion, I think I will leave it as this is.

jksuom commented:
One possible way to simplify the result of `_pair` could be special treatment of the cases where the intervals `(a, b)` and `(c, d)` do not overlap.

    if b <= c:
        br = -oo, -1, a, 0, b, -1, c, 1, d, -1, oo
    elif d <= a:
        br = -oo, -1, c, 1, d, -1, a, 0, b, -1, oo
    else:
        br = ...

I'm not sure if it would help much but it might be worth a try.

smichr commented on sympy/functions/elementary/piecewise.py:
```diff
@@ -227,242 +330,468 @@ def _eval_derivative(self, x):
-        this condition."""
+                a = Dummy('lo')
+                b = Dummy('hi')
+                pos = self._eval_interval(x, a, b, _first=False)
```
Does this look like the right thing to do with Numeric/non-Numeric args?

Also, instead of returning a Piecewise, perhaps a single expression with `sign(hi - lo)*pos` should be returned.

rsemenoff replied:
Hi, I agree that there is a problem with Piecewise itself. It seems nonstandard, and maybe creating unnecessary implementation difficulties? Shouldn't a Piecewise function be just a list of intervals with open boundaries on both sides? And the syntax should require both the upper and lower bound on each and every interval? 
I don't think it unreasonable to expect the user to be completely explicit in defining their input function in this way. 
Or would that entail changes too widespread throughout sympy?

jksuom replied:
I cannot think of a better way for non-numeric args.

The return value should probably be Piecewise. I'm not sure that the single expression would be correct (unless it is Piecewise in disguise: `(sign + 1)*pos/2 + (-sign + 1)*neg/2`).

smichr replied:
I don't know if it's obvious to anyone but changing the order of the limits does not only change the sign of the result, it involves a swap of the symbols representing lower and upper limits. Here is a piece of code taken from the tests added that demonstrates the inequality for a fairly simple expression:

```python
>>> from sympy.utilities.iterables import *
>>> lim = Tuple(x, c, d)
>>> p = Piecewise((1, x < a), (2, x > b), (3, True))
>>> q = p.integrate(lim)
>>> assert q == Piecewise(
...     (-c + 2*d - 2*Min(d, Max(a, c)) + Min(d, Max(a, b, c)), c < d),
...     (-2*c + d + 2*Min(c, Max(a, d)) - Min(c, Max(a, b, d)), True))
>>> pos = q.args[0].expr
>>> eq=Eq(q, sign(d-c)*pos)  # is the general result equivalent to a signed pos value?
>>> for v in permutations((1, 2, 3, 4)):
...     r = dict(zip((a, b, c, d), v))
...     if eq.subs(r) is not S.true:
...         print r
...         break
{c: 3, b: 2, a: 1, d: 4}  # answer: no. These values demonstrate a counter-point.
```

smichr replied:
(In the test suite I exhaustively test the original expression with substituted values and then integrate with the known limits and compare it against the general result.)

jksuom replied:
#13552 turns out to be a good test for piecewise integration.
 
from sympy.stats import *
mixed = Uniform("u1",-1,1) + Uniform("u2",-1,1)
density(mixed)

It seems that uniform densities in sympy.stats are written in terms of Heaviside functions instead of Piecewise expressions. That leads to a rather intractable expression for the density function that consists of two linear pieces: first is increases linearly from 0 to 1/2 in the interval (-2, 0) and then decreases linearly from 1/2 to 0 in the interval (0, 2).

smichr replied:
```
>>> p=Piecewise((0,x<-2),(x/4+S.Half,x<=0),(-x/4+S.Half,x<=2),(0,True))
>>> [p.subs(x,i) for i in range(-3,4)]
[0, 0, 1/4, 1/2, 1/4, 0, 0]
>>> p.integrate(x)
Piecewise((0, x < -2), (x**2/8 + x/2 + 1/2, x <= 0), (-x**2/8 + x/2 + 1/2, x <=2), (1, True))
```

I think I have a step or similar function in the test suite. I'll double check.

Abs.rewrite(Piecewise) should also be defined. The equivalent expression written in terms of Abs is unevaluated:

```
>>> Piecewise((S.Half - abs(x/4), (x>=-2)&(x<=2)), (0,True)).integrate(x)
Piecewise((0, x <= -2), (-(Integral(-2, x) + Integral(Abs(x), x))/4 + Integral(-
2, (x, -2))/4 + Integral(Abs(x), (x, -2))/4, x <= 2), (Integral(-2, (x, -2))/4 -
 Integral(-2, (x, 2))/4 + Integral(Abs(x), (x, -2))/4 - Integral(Abs(x), (x, 2))/4, True))

>>> _.subs(Abs(x),Piecewise((x,x>=0),(-x,True)))
Piecewise((-Piecewise((x, x >= 0), (-x, True))/4 + 1/2, (x >= -2) & (x <= 2)), (0, True))
>>> _.integrate(x)
Piecewise((0, x <= -2), (x**2/8 + x/2 + 1/2, x <= 0), (-x**2/8 + x/2 + 1/2, x <= 2), (1, True))
```

smichr replied:
The triangle-wave shape is integrated in a test at line 205:

```
    g = Piecewise((1 - x, Interval(0, 1).contains(x)),
        (1 + x, Interval(-1, 0).contains(x)), (0, True))
    assert integrate(g, (x, -5, 1)) == 1
    assert integrate(g, (x, 1, -5)) == -1
    assert integrate(g, (x, 1, y)).subs(y, -5) == -1
    assert integrate(g, (x, y, -5)).subs(y, 1) == -1
    i = integrate(g, (x, -5, y))
    assert i.subs(y, 1) == 1
    assert piecewise_fold(i.rewrite(Piecewise)
        ) == Piecewise(
        (1, y >= 1),
        (-y**2/2 + y + 1/2, y >= 0),
        (y**2/2 + y + 1/2, y >= -1),
        (0, True))
    assert i == Piecewise(
        (-Min(-1, y)**2/2 - Min(-1, y) +
            Min(0, y)**2 - Min(1, y)**2/2 + Min(1, y), y > -5),
        (0, True))
    i = integrate(g, (x, y, 1))
    assert i.subs(y, -5) == 1
    assert piecewise_fold(i.rewrite(Piecewise)
        )== Piecewise(
        (1, y <= -1),
        (-y**2/2 - y + 1/2, y <= 0),
        (y**2/2 - y + 1/2, y < 1),
        (0, True))
    assert i == Piecewise(
        (Min(1, Max(0, y))**2 - Min(1, Max(-1, y), Max(0, y))**2/2 -
            Min(1, Max(-1, y), Max(0, y)) + 1/2, y < 1),
        (0, True))
```

This function is defined in a variety of ways (motivated by the original issue). I'm not sure that all those ways as needed...those are slow tests. Perhaps it would be better to leave the variety of conditions to be tested with simpler functions (e.g. `Piecewise((1, cond1),(2, cond2), (0, True))`.

jksuom commented on sympy/functions/elementary/piecewise.py:
```diff
@@ -664,3 +993,210 @@ def piecewise_fold(expr):
+    # neither (-1) as covering that interval and Hi|Lj means
+    # either the hi boundary of the ith interval or the lo
+    # boundary of the jth interval
+    if (b <= c) is S.true:
```
This does not seem to work as expected:
```
>>> (1 <= 2) is S.true
False
```

asmeurer replied:
If you use `is` to compare against S.true, you had better make sure your arguments are sympified. It's more robust to just use `== S.true`.

jksuom commented on sympy/functions/elementary/piecewise.py:
```diff
@@ -1138,25 +1017,24 @@ def _clip(A, B, k):
+    if c != d:
+        p.append((c, d, k))
+    if d != a:
+        p.append((d, a, -1))
```
Do we now have `a <= d <= b`?

smichr replied:
`a <= b` b/c `a, b = Min(a, b), b` at lone 1023; and at 1024, `d = Min(Max(d, a), b)` -- that's not enough for me to see it (though I can slowly reason it out). I test that as:

```
>>> for i in permutations((1,2,3)):
...     a, b, d = i
...     a = min(a, b)
...     d =  Min(Max(d, a), b)
...     assert a <= d and d <= b
```

smichr replied:
When I can use an interactive session I will track down the failures and see what assumptions I've violated. I put the title back to [wip].

smichr committed:
keep only more general br forms

Strunk and White would be proud. Thanks @jksuom for the observation
that this could be removed.

jksuom commented on sympy/functions/elementary/piecewise.py:
```diff
@@ -664,3 +993,234 @@ def piecewise_fold(expr):
+    a, b = B
+    c, d = A
+    A = c, d = Min(Max(c, a), b), Min(Max(d, a), b)
+    p = _pair(A, B, abi=True)
```
I wonder if the previous form could have been used. After

    a, b = B
    c, d = A
    B = a, b = Min(a, b), b
    c, d = Min(Max(c, a), b), Min(Max(d, a), b)
    A = c, d = Min(c, d), d

we should have `a <= c <= d <= b`. Then it would suffice to return those of the three pieces `(a, c, -1)`, `(c, d, k)` and `(d, b, -1)` that are not redundant. Or maybe simply `(a, b, -1)` in case `c == d`.

smichr committed:
remove _pair

It seems like one would be able to remove some of the -1 intervals
from br, but with the removal of duplicates and infinities, I
don't seem to be able to do so without breaking tests.

smichr committed:
remove redundant intervals

smichr commented:
@jksuom , I removed `_pair` and put the pertinent code into `_clip`. Although the use of `a = Min(a, b) and c = Min(c, d)` makes sense, I removed it. Perhaps the way that the Piecewise integration is feeding this routine, that simplification is not necessary -- at least the tests don't complain about not using it. I have alsoremoved what I think are redundant "-1" intervals. I feel like I am [playing Jenga](http://www.jenga.com/about.php) with that piece of code, though, since I don't have a good intuition about it. I'm worried that in removing some of the -1 intervals, some future test will break something.

Can you tell if the ones I've removed are justified?

smichr commented:
holes are handled as follows now:

```
>>> args = (a, Eq(1, x)), (b, And(x > 0, x < 3))
>>> Piecewise(*args).integrate((x,a,b))
Traceback (most recent call last):
...
ValueError: Can't integrate across undefined region.
>>> Piecewise(*args).integrate(x)
Piecewise((nan, x < 0), (b*x, x < 3), (nan, True))
```

smichr commented:
The only thing remaining is to have you take a look at the removed intervals, Kalevi @jksuom and simplify redundant testing as [mentioned above](https://github.com/sympy/sympy/pull/12587#discussion_r148258282). The last TODO is to mark an issue as resolved at stackoverflow.

jksuom commented on sympy/functions/elementary/piecewise.py:
```diff
@@ -1017,43 +1017,41 @@ def _clip(A, B, k):
     a, b = B
     c, d = A
-    a = Min(a, b)
     c, d = Min(Max(c, a), b), Min(Max(d, a), b)
```
This makes  c  and  d  lie between  a  and  b  (assuming that  a <= b), so  b <= c  means that  c == d == b and  d <= a  means that  a == c  == d, and in those cases the  k -interval has length zero.

jksuom commented on sympy/functions/elementary/piecewise.py:
```diff
@@ -1017,43 +1017,41 @@ def _clip(A, B, k):
-        Max(b, c), -1,
-        Max(b, d), -1,
+        Min(c, a), -1,
+        #Min(c, b), -1,
```
It is my understanding that `_clip` should cover only the interval  `(a, b)`. If that is so, it seems to me that we could have  `br = (a, -1, c, k, d, -1, b)` with no (explicit) infinities.

smichr commented:
I investigated to see that `_clip` is only being called with interval bounds generated by `_intervals`; these should return tuples, `t`, where `t[0] <= t[1]`. So that's why the `a = Min(a, b)` and similar for `c` is not needed. Checking your assertion above to be true, I believe the 3 conditions can be reduced to 

```
for i in permutations((1,2,3,4)):
    a,b,c,d=i
    c, d = Min(Max(c, a), b), Min(Max(d, a), b)
    if (b <= c) == True:
        br = (a, -1, c)
    elif (d <= a) == True:
        br = (d, -1, b)
    else:
        br = ((a, -1, c, k, d, -1, b))
```

Let's see how that works. It would no longer require collapsing intervals, either. Nor infinity handling.

smichr committed:
condense code

smichr committed:
simplify piecewise tests

smichr commented:
I was wrong about the `Min(a, b)` being needed; it is now applied only when needed (I think...and existing tests agree): I don't apply it before the contraction, only after.

I'll wait to see if there are any comments before committing. I really appreciate the help and feedback. I think I'm ready for a sabbatical. It's been just over 7 years since I started with SymPy.

smichr commented:
update for max/min `_eval_simplify`
```python
def _eval_simplify(self):
  def opposite(self):
    if not isinstance(self, (Max, Min)):
      return
    return Max if isinstance(self, Min) else Min

  # simplify Min(foo, y, Max(foo, x)) to Min(foo, y)
  # Like And/Or, anything in common between the two
  # is cause to eliminate the opposite arg.
  op = opposite(self)
  if op is None:
    return
  unnested = set()
  nested = []
  remove = []
  for a in self.args:
      if isinstance(a, op):
          nested.append(a)
      else:
          unnested.add(a)
  N = len(nested) - 1
  for i, n in enumerate(reversed(nested)):
      if set(n.args) & unnested:
          nested.pop(N - i)
  rv = self.func(*(list(unnested) + nested))

  # comparable elements allow arg elimination, e.g.
  # Min(1, Max(2, y)) -> Min(1) -> 1 since Max(2, y) >= 2
  # We take only the low hanging fruit here since finding
  # all comparable elements could take a long time; the
  # comparable numbers should be in slot 0
  op = opposite(rv)
  if op is None:
    return rv
  a = rv.args[0]
  numeric = [(a, 0)] if a.is_comparable else []
  for i, a in enumerate(rv.args):
    if isinstance(a, op) and a.args[0].is_comparable:
      numeric.append((a.args[0], i))
  if len(numeric) > 1:
    numeric = list(sorted(numeric))
    if isinstance(rv, Min):
      numeric = list(reversed(numeric))
    numeric.pop()  # this is the one we will keep
    args = list(rv.args)
    for i in sorted([i for _, i in numeric], reverse=True):
      args.pop(i)
    rv = rv.func(*args)

  # similar args that contain a common element can be
  # combined e.g. Min(a, Max(x, y), Max(x, z)) ->
  # Min(a, Max(x, y, z))
  op = opposite(rv)
  if op is None:
    return rv
  m = []
  M = []
  fm = rv.func
  FM = op
  remove = []
  args = list(rv.args)
  for i, a in enumerate(args):
    if isinstance(a, fm):
      m.append(set(a.args))
    elif isinstance(a, FM):
      M.append(set(a.args))
    else:
      continue
    remove.append(i)
  for i in reversed(remove):
    args.pop(i)
  def combine(args):
    hit = True
    while hit:
      hit = False
      for i in range(len(args) - 1):
        for j in range(i + 1, len(args)):
          # https://stackoverflow.com/questions/3170055/
          # test-if-lists-share-any-items-in-python/17735466#17735466
          if not args[i].isdisjoint(args[j])
            args[i] = args[i] | args.pop(j)
            hit = True
            break
        else:
          continue
        break
    return args
  args.extend([fm(*i) for i in combine(m)])
  args.extend([FM(*i) for i in combine(M)])
  return rv.func(*args)
```

moorepants commented:
@smichr Thanks for all your hard work. We hope your sabbatical isn't too long :)

smichr commented:
Thanks, @moorepants!

jksuom commented:
> We hope your sabbatical isn't too long

My feelings too :)

asmeurer commented:
Thank you for your continued work on SymPy @smichr. Your work in this community has been highly valuable, but burnout is real, and we understand if you need to take a break from time to time.

rsemenoff commented:
I agree, the work is valuable and I as well as many others I'm sure are grateful for this important software. One hopes there is a proportionally significant career payoff for the contributors. 
Thanks Richard particularly for getting the merge done, which is allowing me to get back into applied math even though it's been many years since I've looked at it. Take care..
Robert
--------------------------------------------
On Sun, 11/5/17, Aaron Meurer <notifications@github.com> wrote:

 Subject: Re: [sympy/sympy] Piecewise: _eval_interval and _eval_integral overhaul (#12587)
 To: "sympy/sympy" <sympy@noreply.github.com>
 Cc: "rsemenoff" <semenoff@yahoo.com>, "Comment" <comment@noreply.github.com>
 Date: Sunday, November 5, 2017, 5:46 PM
 
 Thank you for your continued work
 on SymPy @smichr. Your work in
 this community has been highly valuable, but burnout is
 real, and we understand if you need to take a break from
 time to time.
 
 —
 You are receiving this because you commented.
 Reply to this email directly, view
 it on GitHub, or mute
 the thread.

certik commented:
@smichr thank you for fixing this, a colleague found a problem reported in #13696, and it turns out that it was already fixed by you in this PR.

Thanks for all your fixes in SymPy over the years, we all really appreciate it.
